File: apps/nextjs-app/src/lib/roles.ts
/**
 * Role extraction — the same rule every app in this repository uses:
 *  - read ROLES_CLAIM as a dotted path from the decoded ACCESS token
 *    ("roles" for the mock IdP / Entra, "realm_access.roles" for Keycloak, "groups" for Oracle),
 *  - the value may be an array of strings or a space-separated string (like `scp`/`scope`),
 *  - the user has app role X when that list contains ROLE_X,
 *  - a missing claim means "authenticated, but no roles".
 * Pure functions (no framework imports) so they are unit-testable with node --test.
 */
import { decodeJwt } from "jose";

export type AppRole = "user" | "admin";

export interface RoleRules {
  rolesClaim: string;
  roleUser: string;
  roleAdmin: string;
}

export type Claims = Record<string, unknown>;

/** Walks a dotted path ("realm_access.roles") through nested objects; undefined when any segment is missing. */
export function readClaimPath(claims: Claims | undefined, path: string): unknown {
  let current: unknown = claims;
  for (const segment of path.split(".")) {
    if (current === null || typeof current !== "object") return undefined;
    current = (current as Claims)[segment];
  }
  return current;
}

/** Raw role/group values found at the configured path, normalised to a string array. */
export function rawRoles(claims: Claims | undefined, rolesClaim: string): string[] {
  const value = readClaimPath(claims, rolesClaim);
  if (Array.isArray(value)) return value.filter((v): v is string => typeof v === "string");
  if (typeof value === "string") return value.split(" ").filter(Boolean);
  return [];
}

/** Maps IdP-specific values to the two application roles. `admin` does NOT imply `user`. */
export function appRoles(claims: Claims | undefined, rules: RoleRules): AppRole[] {
  const raw = rawRoles(claims, rules.rolesClaim);
  const roles: AppRole[] = [];
  if (raw.includes(rules.roleUser)) roles.push("user");
  if (raw.includes(rules.roleAdmin)) roles.push("admin");
  return roles;
}

/** Decodes a JWT payload WITHOUT verifying it; returns undefined for opaque (non-JWT) tokens. */
export function decodeClaims(token: string | undefined): Claims | undefined {
  if (!token) return undefined;
  try {
    return decodeJwt(token) as Claims;
  } catch {
    return undefined;
  }
}

/**
 * Roles for the session: taken from the access token (the artefact the APIs authorise on).
 * The access token came straight from the IdP's token endpoint over a server-to-server call,
 * so decoding it without signature verification is safe here — the APIs still verify it.
 * If the IdP issues opaque access tokens, fall back to the (already validated) ID token claims.
 */
export function rolesFromTokens(accessToken: string, idTokenClaims: Claims | undefined, rules: RoleRules): AppRole[] {
  const access = decodeClaims(accessToken);
  if (access && readClaimPath(access, rules.rolesClaim) !== undefined) return appRoles(access, rules);
  return appRoles(idTokenClaims, rules);
}

File: apps/rest-api/internal/auth/roles.go
// Package auth verifies OIDC access tokens (JWT) and turns their claims into a
// Principal with application roles. The IdP authenticates users and manages
// roles; this package only checks what the IdP signed.
package auth

import "strings"

// Role names used by the application. Which IdP value maps to them is
// configured with ROLE_USER / ROLE_ADMIN (see RoleNames).
const (
	RoleUser  = "user"
	RoleAdmin = "admin"
)

// RolesFromClaims reads the claim at a dotted path (ROLES_CLAIM), e.g.
// "roles", "realm_access.roles", "resource_access.k8sgateway-api.roles" or
// "scp". The value may be a JSON array of strings (Keycloak, Entra roles/groups)
// or one space-separated string (Entra "scp", RFC 9068 "scope"). A missing or
// differently typed claim yields no roles - the caller stays authenticated.
// Limitation: path segments are split on "." so a client id containing dots
// cannot be addressed.
func RolesFromClaims(claims map[string]any, path string) []string {
	var cur any = claims
	for _, part := range strings.Split(path, ".") {
		m, ok := cur.(map[string]any)
		if !ok {
			return nil
		}
		if cur, ok = m[part]; !ok {
			return nil
		}
	}
	switch v := cur.(type) {
	case []any:
		out := make([]string, 0, len(v))
		for _, e := range v {
			if s, ok := e.(string); ok {
				out = append(out, s)
			}
		}
		return out
	case string:
		return strings.Fields(v)
	}
	return nil
}

// RoleNames holds the IdP values that grant the application roles.
type RoleNames struct {
	User  string // ROLE_USER, default "user"
	Admin string // ROLE_ADMIN, default "admin"
}

// AppRoles maps raw IdP values to application roles ("admin", "user").
// Anything else in the claim (Keycloak's default-roles-*, offline_access, ...)
// is ignored.
func (n RoleNames) AppRoles(raw []string) []string {
	roles := make([]string, 0, 2)
	if n.Admin != "" && contains(raw, n.Admin) {
		roles = append(roles, RoleAdmin)
	}
	if n.User != "" && contains(raw, n.User) {
		roles = append(roles, RoleUser)
	}
	return roles
}

func contains(list []string, s string) bool {
	for _, v := range list {
		if v == s {
			return true
		}
	}
	return false
}
